modules/table/src/main/java/org/apache/ignite/internal/table/distributed/index/IndexBuildTask.java

```diff
@@ -0,0 +1,223 @@
+                            LOG.error("Index build error: [{}]", throwable, getCommonIndexInfo());
+
+                            taskFuture.completeExceptionally(throwable);
+                        } else {
+                            taskFuture.complete(null);
+                        }
+                    });
+        } catch (Throwable t) {
+            taskFuture.completeExceptionally(t);
+
+            throw t;
+        } finally {
+            leaveBusy();
+        }
+    }
+
+    /**
+     * Stops index building.
+     */
+    void stop() {
+        if (!taskStopGuard.compareAndSet(false, true)) {
+            return;
+        }
+
+        taskBusyLock.block();
+    }
+
+    /**
+     * Returns the index build future.
+     */
+    CompletableFuture<Void> getTaskFuture() {
+        return taskFuture;
+    }
+
+    private CompletableFuture<Void> handleNextBatch() {
+        if (!enterBusy()) {
+            return completedFuture(null);
+        }
+
+        try {
+            List<RowId> batchRowIds = createBatchRowIds();
+
+            return raftClient.run(createBuildIndexCommand(batchRowIds))
+                    .thenComposeAsync(unused -> {
+                        if (indexStorage.getNextRowIdToBuild() == null) {
+                            // Index has been built.
+                            return completedFuture(null);
+                        }
+
+                        return handleNextBatch();
+                    }, executor);
+        } catch (Throwable t) {
+            return failedFuture(t);
+        } finally {
+            leaveBusy();
+        }
+    }
+
+    private List<RowId> createBatchRowIds() {
+        RowId nextRowIdToBuild = indexStorage.getNextRowIdToBuild();
+
+        List<RowId> batch = new ArrayList<>(batchSize);
+
+        for (int i = 0; i < batchSize && nextRowIdToBuild != null; i++) {
+            nextRowIdToBuild = partitionStorage.closestRowId(nextRowIdToBuild);
+
+            if (nextRowIdToBuild == null) {
+                break;
+            }
+
+            batch.add(nextRowIdToBuild);
+
+            nextRowIdToBuild = nextRowIdToBuild.increment();
+        }
+
+        return batch;
+    }
+
+    private BuildIndexCommand createBuildIndexCommand(List<RowId> rowIds) {
+        boolean finish = rowIds.size() < batchSize;
+
+        return TABLE_MESSAGES_FACTORY.buildIndexCommand()
+                .tablePartitionId(TABLE_MESSAGES_FACTORY.tablePartitionIdMessage()
+                        .tableId(taskId.getTableId())
+                        .partitionId(taskId.getPartitionId())
+                        .build()
+                )
+                .indexId(taskId.getIndexId())
+                .rowIds(rowIds.stream().map(RowId::uuid).collect(toList()))
+                .finish(finish)
+                .build();
+    }
+
+    private boolean enterBusy() {
+        if (!taskBusyLock.enterBusy()) {
+            return false;
+        }
+
+        if (!busyLock.enterBusy()) {
```

Comment:
Wouldn't it be easier to check the `busyLock` first and then the `taskBusyLock`?

Reply:
It doesn't matter to me - I'll change it.